ecloud/resource_volume.go

```diff
@@ -0,0 +1,223 @@
+		Delay:      3 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for volume with ID [%s] to be created: %s", d.Id(), err)
+	}
+
+	return resourceVolumeRead(d, meta)
+}
+
+func resourceVolumeRead(d *schema.ResourceData, meta interface{}) error {
+	service := meta.(ecloudservice.ECloudService)
+
+	log.Printf("[INFO] Retrieving volume with ID [%s]", d.Id())
+	volume, err := service.GetVolume(d.Id())
+	if err != nil {
+		switch err.(type) {
+		case *ecloudservice.VolumeNotFoundError:
+			d.SetId("")
+			return nil
+		default:
+			return err
+		}
+	}
+
+	d.Set("vpc_id", volume.VPCID)
+	d.Set("name", volume.Name)
+	d.Set("capacity", volume.Capacity)
+	d.Set("iops", volume.IOPS)
+
+	return nil
+}
+
+func resourceVolumeUpdate(d *schema.ResourceData, meta interface{}) error {
+	service := meta.(ecloudservice.ECloudService)
+
+	patchReq := ecloudservice.PatchVolumeRequest{}
+	hasChange := false
+	if d.HasChange("name") {
+		hasChange = true
+		patchReq.Name = d.Get("name").(string)
+	}
+	if d.HasChange("capacity") {
+		hasChange = true
+		patchReq.Capacity = d.Get("capacity").(int)
+	}
+	if d.HasChange("iops") {
+		hasChange = true
+		patchReq.IOPS = d.Get("iops").(int)
+	}
+
+	if hasChange {
+		log.Printf("[INFO] Updating volume with ID [%s]", d.Id())
+		taskID, err := service.PatchVolume(d.Id(), patchReq)
+		if err != nil {
+			return fmt.Errorf("Error updating volume with ID [%s]: %w", d.Id(), err)
+		}
+
+		stateConf := &resource.StateChangeConf{
+			Target:     []string{ecloudservice.TaskStatusComplete.String()},
+			Refresh:    VolumeTaskStatusRefreshFunc(service, taskID),
+			Timeout:    d.Timeout(schema.TimeoutUpdate),
+			Delay:      5 * time.Second,
+			MinTimeout: 3 * time.Second,
+		}
+
+		_, err = stateConf.WaitForState()
+		if err != nil {
+			return fmt.Errorf("Error waiting for volume with ID [%s] to return sync status of [%s]: %s", d.Id(), ecloudservice.SyncStatusComplete, err)
+		}
+	}
+	return resourceVolumeRead(d, meta)
+}
+
+func resourceVolumeDelete(d *schema.ResourceData, meta interface{}) error {
+	service := meta.(ecloudservice.ECloudService)
+
+	log.Printf("[INFO] Removing volume with ID [%s]", d.Id())
+	_, err := service.DeleteVolume(d.Id())
+	if err != nil {
+		switch err.(type) {
+		case *ecloudservice.VolumeNotFoundError:
+			return nil
+		default:
+			return fmt.Errorf("Error removing volume with ID [%s]: %s", d.Id(), err)
+		}
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Refresh:    VolumeNotFoundRefreshFunc(service, d.Id()),
+		Timeout:    d.Timeout(schema.TimeoutDelete),
+		Delay:      5 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for volume with ID [%s] to be deleted: %s", d.Id(), err)
+	}
+
+	return nil
+}
+
+func VolumeTaskStatusRefreshFunc(service ecloudservice.ECloudService, taskID string) resource.StateRefreshFunc {
```

Comment:
As this can be reused by other resources, it's probably worth making this generic - `TaskStatusRefreshFunc`